### DISCUSSION TITLE
Matter & common bluetooth sensors

### DISCUSSION THREAD

**Discussion by casanoe** [arendst/Tasmota#20388] [Q&A] [answered]
Hi all,
I'm new to the Tasmota firmware, and I'm very impressed with the work that's been done. First of all, thank you all for it.
The reason why I decided to use Tasmota is because I would like to add in my HomeKit environment some HomeKit incompatible sensors I have like LYWSD03MMC (unencrypted). 
So the Matter + Bluetooth LE seems to be the right choice. Unfortunately it doesn't work for me. Here are some more information :
- Custom Firmware compiled with Platformio
[platformio_override.txt](https://github.com/arendst/Tasmota/files/13821893/platformio_override.txt)
[user_config_override.txt](https://github.com/arendst/Tasmota/files/13821894/user_config_override.txt)
- Tasmota 13.3.0.1(TasmoCompiler-esp32generic)
- ESP32 WROVER-E
- Boot : 
```
00:00:00.416 MTR: Loaded 2 fabric(s)
00:00:00.438 BRY: Berry initialized, RAM used 20471 bytes
00:00:00.460 Projet tasmota - Tasmota Version 13.3.0.1(TasmoCompiler-esp32generic)-2_0_14(2024-01-03T10:48:01)
00:00:00.461 iBeacon register for advert callbacks
00:00:00.472 MI32Option0: 1
00:00:00.473 MI32Option1: 0
00:00:00.473 MI32Option2: 0
00:00:00.474 MI32Option4: 1
00:00:00.474 MI32Option5: 0
00:00:00.485 MI32Option6: 0
00:00:00.485 M32: init: request callbacks
00:00:04.307 MTR: Configuring endpoints
00:00:04.310 MTR:   endpoint =     0 type:root
00:00:04.320 MTR:   endpoint =     1 type:temperature filter:ATC863cc1#Temperature
00:00:04.325 MTR:   endpoint =     2 type:v_temp name:test
00:00:04.327 MTR:   endpoint = 65280 type:aggregator
```
- The sensor LYWSD03MMC is correctly discovered by the Bluetooth
`SENSOR = {"Time":"2024-01-03T18:46:15","ATC863cc1":{"mac":"a4c138863cc1","Temperature":20.2,"Humidity":55.0,"DewPoint":10.9,"Battery":100,"RSSI":-41},"TempUnit":"C"}`
- The sensor LYWSD03MMC is correctly published in HomeKit but returns always 0°C
- I configured the sensor in Matter configuration page like this : type = Temperature, Parameter = ATC863cc1#Temperature
- Result of MTRINFO : (Ep 1 is a real LYWSD03MMC, Ep 2 is a virtual sensor)
```
18:47:02.871 RSL: RESULT = {"MtrInfo":{"Ep":1,"Temperature":null}}
18:47:02.881 RSL: RESULT = {"MtrInfo":{"Ep":2,"Name":"test","Temperature":null}}
18:47:02.889 RSL: RESULT = {"MtrInfo":"Done"}
```
- I "MTRUPDATE" the virtual sensor Ep 2 like this, but the result in HomeKit is strange. It returns 0.5°C
```
18:55:01.921 CMD: mtrupdate {"Ep":2,"temperature":66}
18:55:01.939 RSL: RESULT = {"MtrUpdate":{"Ep":2,"Name":"test","Temperature":66}}
```

What's wrong? Can you help me? Thank you very much!

**Comment by s-hadinger** [comment]
The temperature in Matter is in 1/100 of degrees. You need to multiply by 100. 

The value 66 is 0.66 degrees rounded to 0.5

**Reply by casanoe** [reply]
Great. Thanks a lot. It works fine with the virtual sensor, using Mtrupdate command. But I don’t know how to change the value (*100) of the Bluetooth sensor on the fly

**Reply by s-hadinger** [reply]
The multiplication is only needed for `MtrUpdate`. It is automatically done for sensors. Can you set the logs to `weblog 3` and capture when a temperature change occurs?

**Reply by casanoe** [reply]
Here is : 
```
10:13:52.929 MTR: <Sub_Alive ( 49276) sub=10554
10:13:54.661 MTR: read_sensors: nil
10:13:59.675 MTR: read_sensors: nil
10:14:04.683 MTR: read_sensors: nil
10:14:09.922 MTR: read_sensors: nil
10:14:14.939 MTR: read_sensors: nil
10:14:19.949 MTR: read_sensors: nil
10:14:25.165 MTR: read_sensors: nil
10:14:30.179 MTR: read_sensors: nil
10:14:35.187 MTR: read_sensors: nil
10:14:40.415 MTR: read_sensors: nil
10:14:45.447 MTR: read_sensors: nil
```

**Reply by casanoe** [reply]
and the result of the sensor :
`10:08:03.364 RSL: SENSOR = {"Time":"2024-01-04T10:08:03","ATC863cc1":{"mac":"a4c138863cc1","Temperature":19.4,"Humidity":56.0,"DewPoint":10.4,"Battery":100,"RSSI":-53},"TempUnit":"C"}`

**Reply by s-hadinger** [reply]
Oh, I see. BLE are not regular sensors. Matter normally uses `Status 8` to read the value of sensors. But in the case of BLE, they are  sent once received and not included in `Status 8`.

So you need to define the Matter endpoint as virtual (no pattern) and then setup a rule to get the value from the BLE sensor, and do `MtrUpdate` with the value multiplied by 100.

**Reply by sfromis** [reply, marked as answer]
Reasonably straightforward with a Berry one-liner like:
`tasmota.add_rule('ATC863cc1#Temperature', /value->tasmota.cmd(f'MtrUpdate {{"Ep":2,"Temperature":{value*100}}}')`

**Comment by barbudor** [comment]
> Oh, I see. BLE are not regular sensors. Matter normally uses `Status 8` to read the value of sensors. But in the case of BLE, they are sent once received and not included in `Status 8`.

Isn't that the same with Zigbee ?
If you are doing better in Zigbee, may be there would be something to backport from ZB to BLE

**Reply by sfromis** [reply]
Like BLE, Zigbee is not "regular" sensors either, but extra messages outside of the normal Tasmota sensor framework, and therefore not appearing in Status 10/8 and discovery payloads.

**Reply by barbudor** [reply]
that what I thought, just wanted to be sure

**Reply by casanoe** [reply]
Thanks a lot for your advises and explanation. It’s clear. I started wtiting a berry driver to  do the job automatically. I will share it as soon as it’s ready.

**Reply by casanoe** [reply]
Here is a driver ("quick & dirty"). It works for me (3 LYWSD03MMC, publishing Temp & Hum):
```
#-
 - Driver written in Berry
 - Automaticaly plublish MI32 sensors in Matter
 -#

 import global
 global.ATCDriver_value = {}
 
 class ATCDriver
   var ok
   var last_updated
   var seconds
 
   def init()
     tasmota.cmd("MI32OPTION2 1")
     log("ATCDriver: Init...")
     self.last_updated = 0
     self.seconds = 0
     self.check(true)
   end
 
   def update_mtr_devices(b)
     import path
     var dt = path.last_modified("_matter_device.json")
     if (dt > self.last_updated) || b
       import json
       import string
       log("ATCDriver: Update devices", 3)
       global.ATCDriver_value = {}
       self.last_updated= tasmota.time_dump(tasmota.rtc()['local'])['epoch']
       var f = open("_matter_device.json")
       var d1 = json.load(f.read())['config']
       f.close()
       var se = json.load(tasmota.read_sensors())
       var dev
       self.remove_allrules()
       for v:d1
         if v.find('filter')
           dev = string.split(v['filter'], '#')[0]
           if global.ATCDriver_value.find(dev) == nil && se.find(dev) == nil
             self.add_rule(dev)
             global.ATCDriver_value[dev] = {}
           end
         end
       end
     end
   end
 
   def add_rule(dev)
     log("ATCDriver: Add rule "..dev, 3)
     tasmota.add_rule(dev, 
       def(value) 
         import global
         log("ATCDriver: update "..dev, 3)
         global.ATCDriver_value[dev] = value
       end
       , "ATCDriver_"..dev)
     end

   def remove_allrules()
     import string
     if tasmota._rules
       var i = 0
       while i < tasmota._rules.size()
         if string.find(tasmota._rules[i].id, 'ATCDriver_') == 0
           log("ATCDriver: remove rule "..str(i).." "..tasmota._rules[i].id, 3)
           tasmota._rules.remove(i)
         else
           i += 1
         end
       end
     end
   end
 
   def check(b)
     import path
     log("ATCDriver:  check", 3)
     if path.exists("_matter_device.json")
       self.ok = true
       self.update_mtr_devices(b)
     end
   end
 
   #- trigger a read every second -#
   def every_second()
     self.seconds = self.seconds + 1
     if self.seconds == 60
       self.check()
       self.seconds = 0
     end
   end
 
   #- display sensor value in the web UI -#
   def web_sensor()
     if !self.ok return nil end  #- exit if not initialized -#
     var msg = "{s}{m}{e}"..
      "{s}ATCDriver {m}"..str(size(global.ATCDriver_value)).." devices{e}"..
      "{s}ATCDriver update in{m}"..str(60-self.seconds).." seconds{e}"
     tasmota.web_send_decimal(msg)
   end
 
   #- add sensor value to teleperiod -#
   def json_append()
     if !self.ok return nil end  #- exit if not initialized -#
     import global
     import json
     import string
     for dev:global.ATCDriver_value.keys()
       var v = global.ATCDriver_value[dev]
       v = string.replace(string.format("%s",v), "'", "\"")
       #print("ATCDriver: val == "..v)
       tasmota.response_append(',"'..dev..'":'..v)
       log("ATCDriver: response append "..dev, 3)
     end
   end
 
 end
 
 atc = ATCDriver()
 tasmota.add_driver(atc)

```

Every 60s the driver checks if Matter endpoints changed. 
I haven't done much testing. It's possible that there are still bugs.

**Reply by casanoe** [reply]
Script hosted & updated here : https://github.com/casanoe/Tasmota_Berry/tree/main
